torch/csrc/jit/codegen/cuda/fusion.cpp

```diff
@@ -601,6 +612,55 @@ std::vector<Val*> Fusion::getTerminatingOutputs() {
   return terminating_outputs;
 }
 
+void Fusion::aliasOutputToInput(Val* output, Val* input) {
+  TORCH_INTERNAL_ASSERT(
```

Comment:
I think the check here should be if it has the output, then it must have the input. Not that they should both always be here. Also should add a comment that this is due to segmentation where we may use the input, and we could alias it, but if we don't produce the output then aliasing doesn't have to happen.

Reply:
We should also clean up the returned alias if entries aren't in both input and output for a given alias.

Reply:
Since remove/add/output/input can be called multiple times on a fusion, we should validate at lowering that if an output in the alias map is there, then the corresponding input must be registered.

Reply:
> I think the check here should be if it has the output, then it must have the input. Not that they should both always be here.

This API is used in parser. So at the time when we are defining the computation, segmentation hasn't happened yet. Hence the check here is just a sanity check. Because the only thing we can support via integration is to alias from an output to an input. I agree the other check is a lot more important, but this one also confines the scope of our aliasing support.

> add a comment that this is due to segmentation where we may use the input, and we could alias it, but if we don't produce the output then aliasing doesn't have to happen.

I'll put a comment as well as a TODO. Will work with Shiming to get this done.

> Since remove/add/output/input can be called multiple times on a fusion, we should validate at lowering that if an output in the alias map is there, then the corresponding input must be registered.

Sounds good. I'll add that check.

Reply:
Opened #898 to track this